### PR TITLE
[Feature] Support sequence parallelism

### PR DESCRIPTION
**Pull request by cascade812**
support sequence parallel with TP on models like llama

In this PR, I modified RowParallelLinear, ColumnParallelLinear, LogitsProcessor, VocabParallelEmbedding to support SP. 

Belows are TODOs
- support combination with pp
- support other layers

### CODE REVIEW

**Comment by yaochengji** [comment]
Thanks for your contribution, @cascade812 !

Hi @robertgshaw2-redhat , @tlrmchlsmth , could you take a look at this?

Based on my micro-benchmark, collective-matmul optimization can improve the performance of multi-chip inference on TPU greatly. To enable collective-matmul, we depend on vLLM to support megatron-style sequence parallelism. Then the TPU compiler can automatically convert the pattern of ag-matmul and matmul-rs to collective-matmul.

cc @bvrockwell @yarongmu-google

**Comment by tlrmchlsmth** [comment]
I’ll take a look! I’m excited about and in favor of sequence parallel support in general.

@yaochengji could you explain why this helps the gemm-rs and ag-gemm rewrite? I don’t really see that as sequence parallel outside of the rms_norm

**Comment by yaochengji** [comment]
Thanks, @tlrmchlsmth .

It is because sequence parallelism, the model with tp looks like as below:
matmul -> allreduce -> rms_norm(or other ops) -> matmul

with sp enabled, it becomes:
matmul -> rs -> rms_norm(or other ops) -> ag -> matmul

TPU compiler can only detect the matmul-rs and ag-matmul patterns and replace them with corresponding optimized collective-matmul. And I believe the simple pattern with sp enabled also benefits other hardware.

**Review by youkaichao — COMMENTED** [review]
our previous plan for sequence parallel, is to make it a compilation pass, without changing the linear/embedding layer. that thread somehow got lost.

cc @bnellnm

**Comment by yaochengji** [comment]
Thanks for your comment, @youkaichao !

If we don't change the linear/embedding layer, we have to match various patterns for `matmul -> allreduce -> rms_norm(or other ops) -> matmul` even when only rms_norm is considered. We might find different kinds of hardware have different rms_norm implementation, sometimes one hardware has various implementations of rms_norm, e.g. [nvgpu](https://github.com/vllm-project/vllm/blob/main/vllm/model_executor/layers/layernorm.py#L76).

Changing the linear or embedding layer offers a potentially more generalized approach. Furthermore, the number of layers requiring modification is relatively small.

**Comment by yaochengji** [comment]
@cascade812 I know currently not all layers support sp. Could you print readable message to the vLLM users if a specific layer doesn't support sequence parallelism?

**Review by NickLucche — CHANGES_REQUESTED** [review]
Mega job here thanks!
Can we get some benchmark numbers to go along with this PR?

**Comment by bnellnm** [comment]
> our previous plan for sequence parallel, is to make it a compilation pass, without changing the linear/embedding layer. that thread somehow got lost.
> 
> cc @bnellnm

There were too many problems with pytorch, the limitations of the kernels and issues with piecewise graphs so the project was put on hold.

**Comment by yaochengji** [comment]
> Mega job here thanks! Can we get some benchmark numbers to go along with this PR?

I'm working on enabling collective matmul optimization on TPU. My change on vLLM will be based on this PR, will share the benchmark numbers later.

File: vllm/distributed/parallel_state.py [hunk 1]
```diff
@@ -312,6 +312,13 @@ def all_reduce(self, input_: torch.Tensor) -> torch.Tensor:
     def _all_reduce_out_place(self, input_: torch.Tensor) -> torch.Tensor:
         return self.device_communicator.all_reduce(input_)
 
+    def reduce_scatter(self, input_: torch.Tensor) -> torch.Tensor:
+        # Bypass the function if we are using only 1 GPU.
+        if self.world_size == 1:
+            return input_
+
+        return self.device_communicator.reduce_scatter(input_)
```
Comment (tlrmchlsmth) [inline vllm/distributed/parallel_state.py:320]: reduce_scatter should take a `dim`, just like `all_gather` does (I have some code here you can steal do handle this https://github.com/vllm-project/vllm/pull/14057)

File: vllm/model_executor/layers/linear.py [hunk 2]
```diff
@@ -1258,8 +1265,14 @@ def forward(
         output_parallel = self.quant_method.apply(self,
                                                   input_parallel,
                                                   bias=bias_)
+
         if self.reduce_results and self.tp_size > 1:
-            output = tensor_model_parallel_all_reduce(output_parallel)
+            forward_context = try_get_forward_context()
+            if (forward_context is not None
+                    and forward_context.enable_sequence_parallel):
```
Comment (tlrmchlsmth) [inline vllm/model_executor/layers/linear.py:1272]: The forward context isn't available outside of model initialization, so you'll have to do  `self.enable_sequence_parallel = forward_context.enable_sequence_parallel` in `__init__`, otherwise you won't actually be using sequence parallel while actually inferencing the model (unless you're using CUDA graphs)

I think this is a pretty tricky footgun, so we should address this - (cc @youkaichao)

File: vllm/model_executor/models/llama.py
Comment (tlrmchlsmth) [inline vllm/model_executor/models/llama.py:1]: Please revert the spurious changes to this file, thx

File: vllm/v1/worker/gpu_model_runner.py [hunk 3]
```diff
@@ -1295,9 +1320,17 @@ def _dummy_run(
                     for k, v in self.intermediate_tensors.items()
                 })
 
-            with set_forward_context(None,
-                                     self.vllm_config,
-                                     num_tokens=num_tokens):
+            enable_sequence_parallel = (
+                self.vllm_config.parallel_config.enable_sequence_parallel
+                and num_tokens %
+                self.vllm_config.parallel_config.tensor_parallel_size == 0)
```
Comment (tlrmchlsmth) [inline vllm/v1/worker/gpu_model_runner.py:1326]: Decisions like this should go in `vllm/config.py`

File: vllm/v1/worker/gpu_model_runner.py [hunk 4]
```diff
@@ -1041,7 +1053,11 @@ def execute_model(
 
         hidden_states = hidden_states[:num_scheduled_tokens]
         sample_hidden_states = hidden_states[logits_indices]
-        logits = self.model.compute_logits(sample_hidden_states, None)
+        with set_forward_context(
+                attn_metadata,
+                self.vllm_config,
+                enable_sequence_parallel=enable_sequence_parallel):
```
Comment (tlrmchlsmth) [inline vllm/v1/worker/gpu_model_runner.py:1059]: OK, I see we're setting the forward_context here now. @youkaichao thoughts on this?

**Review by tlrmchlsmth — COMMENTED** [review]
I left some inline comments. Generally I also think this should be done as a pass in torch inductor or similar compiler layer. I'm pretty sure these changes are making assumptions about the model definition that may not be valid.

**Comment by cascade812** [comment]
Thanks for all the comments and reviews! I'll address accordingly. 

> I know currently not all layers support sp. Could you print readable message to the vLLM users if a specific layer doesn't support sequence parallelism?

@yaochengji  Do you mean adding a message to all unsupported layers? That would involve many layers. Do you have any suggestions on how to implement this efficiently?

File: vllm/distributed/parallel_state.py (hunk 1 above)
Comment (cascade812) [reply vllm/distributed/parallel_state.py:320]: sure, will update

File: vllm/model_executor/layers/linear.py (hunk 2 above)
Comment (cascade812) [reply vllm/model_executor/layers/linear.py:1272]: The forward context is available, it's set before calling self.model().

File: vllm/v1/worker/gpu_model_runner.py (hunk 3 above)
Comment (cascade812) [reply vllm/v1/worker/gpu_model_runner.py:1326]: This logic is placed outside of vllm.config because torch.distributed.reduce_scatter_tensor only works when the reduced scattered dimension (in this case, the token dimension) is divisible by the parallel size. Since num_tokens changes every iteration, it seems not reasonable to put it in the config.

**Comment by yaochengji** [comment]
> Do you mean adding a message to all unsupported layers? That would involve many layers. Do you have any suggestions on how to implement this efficiently?

Usually we have a base class and the sequence parallelism `not implemented warning` can be put there. But I just took a look the vLLM code realize it's not easy to implement that way. 

NVM, I'm fine that users need to be aware of the sequence parallelism support availability before it is fully supported.

**Comment by youkaichao** [comment]
> we have to match various patterns for matmul -> allreduce -> rms_norm(or other ops) -> matmul even when only rms_norm is considered.

I think you only need to match `matmul -> allreduce` and `allreduce -> matmul` ? `rms_norm` just takes an input and produce output with the same shape, you don't need to change the op.

File: vllm/v1/worker/gpu_model_runner.py (hunk 4 above)
Comment (youkaichao) [reply vllm/v1/worker/gpu_model_runner.py]: looks quite intrusive to me 👀

**Comment by yaochengji** [comment]
> I think you only need to match `matmul -> allreduce` and `allreduce -> matmul` ? 

There's no `allreduce->matmul` in 1D TP. 

> `rms_norm` just takes an input and produce output with the same shape, you don't need to change the op.
We don't need to change the op but we still need to detect it. Because we will decompose `allreduce` to `reduce-scatter` and `allgather` and move `allgather` after `rms_norm` but before the next `matmul`.

File: vllm/v1/worker/gpu_model_runner.py (hunk 4 above)
Comment (cascade812) [reply vllm/v1/worker/gpu_model_runner.py]: @youkaichao  Since we need num_tokens to make the decision before reduce-scatter (see the explanation above), do you have any suggestions for a better approach?

**Comment by cascade812** [comment]
> hey @cascade812 - is it okay if I push a few changes to your branch?

Absolutely! Your contributions are more than welcome. Thanks!

**Comment by cascade812** [comment]
@robertgshaw2-redhat @tlrmchlsmth @youkaichao what's your thought on this? If you guys think compilation pass is the better choice, I'm willing to help too!

**Comment by tlrmchlsmth** [comment]
> @robertgshaw2-redhat @tlrmchlsmth @youkaichao what's your thought on this? If you guys think compilation pass is the better choice, I'm willing to help too!

I didn’t see this message! It would be awesome for you to work on that!

**Comment by cascade812** [comment]
> > @robertgshaw2-redhat @tlrmchlsmth @youkaichao what's your thought on this? If you guys think compilation pass is the better choice, I'm willing to help too!
> 
> I didn’t see this message! It would be awesome for you to work on that!
Great! I'll work on it and keep you posted on the progress.

**Comment by tlrmchlsmth** [comment]
> > > @robertgshaw2-redhat @tlrmchlsmth @youkaichao what's your thought on this? If you guys think compilation pass is the better choice, I'm willing to help too!
> > 
> > 
> > I didn’t see this message! It would be awesome for you to work on that!
> 
> Great! I'll work on it and keep you posted on the progress.

Hey @cascade812, I have some thoughts/pointers on this (sorry, meant to send these along earlier):

First, here is an earlier PR that attempted to do this via an inductor pass with pattern matching: https://github.com/vllm-project/vllm/pull/9886.
There are a couple of issues with this implementation. It has some deadlocking issues that aren't that well understood, and it also has some complexities when working with `num_tokens % 4 != 0` (similar to things that you need to deal with in this PR).

Another problem is that the pattern matching is very brittle and would need to be extended to support different models. @yaochengji raised this issue, and from my past experience I think this is a very valid concern. 

In an effort to make it more flexible, a some of us have discussed adding sentinel no-op operations (e.g. a `begin_sp_region` operation that does a clone at the end of RowParallelLinear's apply method and an `end_sp_region` at the beginning of MergedColumnParallelLinear). Then we can find regions fenced by these operations to do a rewrite. This would let us be more robust and selective, and would neatly handle PP as well

BTW are you on the vllm slack? If not, please join as it's easier to discuss there!

**Comment by yaochengji** [comment]
@tlrmchlsmth 

I noticed that the layer_norm is a `CustomOp` class, do you think we can make it a composite `rms_norm` op in fx graph for all cases?

https://github.com/vllm-project/vllm/blob/db9dfcfa6a0b88fb880ee21b56f133c9c5a600ab/vllm/model_executor/layers/layernorm.py#L81

**Comment by tlrmchlsmth** [comment]
> @tlrmchlsmth
> 
> I noticed that the layer_norm is a `CustomOp` class, do you think we can make it a composite `rms_norm` op in fx graph for all cases?
> 
> https://github.com/vllm-project/vllm/blob/db9dfcfa6a0b88fb880ee21b56f133c9c5a600ab/vllm/model_executor/layers/layernorm.py#L81

Not sure about this. Often we make things custom ops when torch.compile has trouble with them. @youkaichao do you know?

File: vllm/worker/model_runner.py [hunk 5]
```diff
@@ -1736,9 +1736,24 @@ def execute_model(
...
+
+        # only do sequence parallelism when num of tokens
+        # is divisible by parallel size.
+        # sequence parallelism uses torch.distributed.reduce_scatter which only
+        # supports the case when size is divisible by parallel size
+        enable_sequence_parallel = (
+            self.vllm_config.parallel_config.enable_sequence_parallel
+            and num_tokens is not None and num_tokens %
+            self.vllm_config.parallel_config.tensor_parallel_size == 0)
+
```
Comment (mklasby) [inline vllm/worker/model_runner.py:1750]: It may be beneficial to log a warning the first time this occurs so that it's clear in logs if sequence parallel is not being enabled due to sequence length even if config argument for sequence parallel is True.

**Comment by tlrmchlsmth** [comment]
Closing in favor of #16155, which has been merged!
